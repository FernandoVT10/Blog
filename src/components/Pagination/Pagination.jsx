import { useRouter } from "next/router";

import "./Pagination.scss";

export default ({ pagination }) => {
    const router = useRouter();

    const handleOffset = (offset) => {
        const query = router.query;

        Object.assign(query, { offset });

        router.push({
            pathname: router.pathname,
            query
        })
    };

    const getLeftArrow = () => {
        if(pagination.hasPrevPage) {
            return (
                <a
                href="#"
                onClick={() => handleOffset(pagination.prevPage)}
                className="pagination__arrow">
                    <i className="fas fa-chevron-left" aria-hidden="true"></i>
                </a>
            );
        }

        return (
            <a href="#" className="pagination__arrow pagination__arrow--disabled">
                <i className="fas fa-chevron-left" aria-hidden="true"></i>
            </a>
        );
    };

    const getRightArrow = () => {
        if(pagination.hasNextPage) {
            return (
                <a
                href="#"
                onClick={() => handleOffset(pagination.nextPage)}
                className="pagination__arrow">
                    <i className="fas fa-chevron-right" aria-hidden="true"></i>
                </a>
            );
        }

        return (
            <a href="#" className="pagination__arrow pagination__arrow--disabled">
                <i className="fas fa-chevron-right" aria-hidden="true"></i>
            </a>
        );
    };

    const getPages = () => {
        if(pagination.pages) {
            return pagination.pages.map(({ number, active }, index) => {
                const numberClass = active ? "pagination__number--active" : "";

                return (
                    <li className="pagination__number-item" key={index}>
                        <a 
                        href="#"
                        onClick={() => handleOffset(number)}
                        className={`pagination__number ${numberClass}`}>
                            { number }
                        </a>
                    </li>
                );
            });
        }
    }

    return (
        <div className="pagination">
            { getLeftArrow() }

            <ul className="pagination__number-container">
                { getPages() }
            </ul>

            { getRightArrow() }
        </div>
    );
};