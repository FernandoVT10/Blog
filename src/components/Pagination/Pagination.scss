.pagination {
    display: flex;
    justify-content: center;
    padding: 10px 0;
    width: 100%;
    background: var(--primary);

    &__arrow {
        color: var(--white);
        margin: 10px;
        transition: .3s;

        &:hover {
            color: var(--secondary);
        }

        &--disabled,
        &--disabled:hover {
            cursor: default;
            color: var(--gray-light);
        }
    }

    &__number-container {
        margin: 0;
        padding: 0;
    }

    &__number-item {
        display: inline-block;
        padding: 10px;
    }

    &__number {
        color: var(--white);
        font-weight: bold;
        transition: .3s;

        &:hover,
        &--active {
            color: var(--secondary);
        }
    }
}