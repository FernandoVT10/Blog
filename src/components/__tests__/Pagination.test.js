import Pagination from "../Pagination/";
import { create } from "react-test-renderer";
import { useRouter } from "next/router";

const PAGIANATION_MOCK = {
    hasPrevPage: true,
    prevPage: 5,
    hasNextPage: true,
    nextPage: 7,
    pages: [
        { number: 2, active: false },
        { number: 3, active: false },
        { number: 4, active: false },
        { number: 5, active: false },
        { number: 6, active: true },
        { number: 7, active: false },
        { number: 8, active: false }
    ]
};

jest.mock("next/router", () => ({
    useRouter: jest.fn(() => ({ })),
}));

describe("<Pagination/> Component", () => {
    it("Check if renders correctly", () => {
        const component = create(<Pagination pagination={PAGIANATION_MOCK}/>);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("It should redirect to 2 page", () => {
        let routerPush = jest.fn();

        useRouter.mockImplementation(() => ({
            pathname: "/",
            push: routerPush,
            query: {}
        }));

        const component = create(<Pagination pagination={PAGIANATION_MOCK}/>);

        const numbersItem = component.root.findAllByProps({ className: "pagination__number " });

        numbersItem[0].props.onClick();

        expect(routerPush).toHaveBeenCalledWith({
            pathname: "/",
            query: {
                offset: 2
            }
        });
    });

    it("It should redirect to previous page", () => {
        let routerPush = jest.fn();

        useRouter.mockImplementation(() => ({
            pathname: "/",
            push: routerPush,
            query: {}
        }));

        const component = create(<Pagination pagination={PAGIANATION_MOCK}/>);

        const numbersItem = component.root.findByProps({ className: "fas fa-chevron-left" });

        numbersItem.parent.props.onClick();

        expect(routerPush).toHaveBeenCalledWith({
            pathname: "/",
            query: {
                offset: 5
            }
        });
    });

    it("It should redirect to next page", () => {
        let routerPush = jest.fn();

        useRouter.mockImplementation(() => ({
            pathname: "/",
            push: routerPush,
            query: {}
        }));

        const component = create(<Pagination pagination={PAGIANATION_MOCK}/>);

        const numbersItem = component.root.findByProps({ className: "fas fa-chevron-right" });

        numbersItem.parent.props.onClick();

        expect(routerPush).toHaveBeenCalledWith({
            pathname: "/",
            query: {
                offset: 7
            }
        });
    });
});

