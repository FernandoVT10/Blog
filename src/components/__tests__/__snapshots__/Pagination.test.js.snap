// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> Component Check if renders correctly 1`] = `
<div
  className="pagination"
>
  <a
    className="pagination__arrow"
    href="#"
    onClick={[Function]}
  >
    <i
      aria-hidden="true"
      className="fas fa-chevron-left"
    />
  </a>
  <ul
    className="pagination__number-container"
  >
    <li
      className="pagination__number-item"
    >
      <a
        className="pagination__number "
        href="#"
        onClick={[Function]}
      >
        2
      </a>
    </li>
    <li
      className="pagination__number-item"
    >
      <a
        className="pagination__number "
        href="#"
        onClick={[Function]}
      >
        3
      </a>
    </li>
    <li
      className="pagination__number-item"
    >
      <a
        className="pagination__number "
        href="#"
        onClick={[Function]}
      >
        4
      </a>
    </li>
    <li
      className="pagination__number-item"
    >
      <a
        className="pagination__number "
        href="#"
        onClick={[Function]}
      >
        5
      </a>
    </li>
    <li
      className="pagination__number-item"
    >
      <a
        className="pagination__number pagination__number--active"
        href="#"
        onClick={[Function]}
      >
        6
      </a>
    </li>
    <li
      className="pagination__number-item"
    >
      <a
        className="pagination__number "
        href="#"
        onClick={[Function]}
      >
        7
      </a>
    </li>
    <li
      className="pagination__number-item"
    >
      <a
        className="pagination__number "
        href="#"
        onClick={[Function]}
      >
        8
      </a>
    </li>
  </ul>
  <a
    className="pagination__arrow"
    href="#"
    onClick={[Function]}
  >
    <i
      aria-hidden="true"
      className="fas fa-chevron-right"
    />
  </a>
</div>
`;
